### PR TITLE
perf: Optimize the implementation of dflate and add a test

### PR DESCRIPTION
**Pull request by frankdavid**
NODE-1454

The following optimizations were done:
- Use a BufReader and remove unnecessary seeking
- Use a pre-allocated buffer instead of allocating a new buffer for every 512 bytes read
- Modify `is_empty` implementation so Rust generates more efficient assembly

I also added a test which passes both before and after the optimizations.

The optimizations make dflate on average ~2.5x faster in our build invocations, saving ~20seconds when building` //ic-os/guestos/envs/dev`

[NODE-1454]: https://dfinity.atlassian.net/browse/NODE-1454?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by Bownairo — APPROVED** [review]
Awesome!

---

Do you have any ideas why we don't see a similar speedup for HostOS? Their sizes are around the same, but HostOS is way less dense.

`bazel build --config=local ic-os/hostos/envs/dev`

Before:
```
INFO: Elapsed time: 298.408s, Critical Path: 260.33s
INFO: 801 processes: 311 internal, 468 linux-sandbox, 22 local.
INFO: Build completed successfully, 801 total actions
```

After:
```
INFO: Elapsed time: 294.682s, Critical Path: 257.41s
INFO: 801 processes: 311 internal, 468 linux-sandbox, 22 local.
INFO: Build completed successfully, 801 total actions
```

File: rs/ic_os/dflate/src/lib.rs [hunk 1]
```diff
@@ -166,10 +172,11 @@ where
 
 // null check from tar-rs
 fn is_empty(block: &[u8]) -> bool {
-    block.iter().all(|i| *i == 0)
+    // Efficiently check whether all elements are 0.
+    block == &[0; MAX_BLOCK_SIZE][..block.len()]
 }
 
-fn seek_hole(file: &File, from: u64) -> u64 {
+fn seek_hole(file: &mut File, from: u64) -> u64 {
```
[the remark this answers is not in the document]
Comment (frankdavid) [reply rs/ic_os/dflate/src/lib.rs:179]: That's right. File::seek() requires &mut self so our custom implementation should also take &mut (it's not nice to seek a file while someone is holding a reference to it).

**Comment by frankdavid** [comment]
I'm still seeing improvements for the HostOS, albeit not as much (~1.5x speed). There are also less invocations to dflate during the HostOS build (11 vs 13). I didn't analyze the layout of the files but when you say less dense, can it be that most of the work is done by seeking with SEEK_HOLE and SEEK_DATA so we have to crunch through less zeros?
Also note that measuring the entire build time has a lot of noise since most of the time will be spent compiling rust code and other actions, the wall time of which can vary a lot based on the load on the servers that host the build env. Therefore I measured the dflate invocations separately.

**Comment by Bownairo** [comment]
> Therefore I measured the dflate invocations separately.

Ahhh got it, I assumed you were measuring overall, glad to see there's also an improvement here!

> I didn't analyze the layout of the files but when you say less dense, can it be that most of the work is done by seeking with SEEK_HOLE and SEEK_DATA so we have to crunch through less zeros?

That's what I was thinking, but my measurement with almost no difference was throwing me off. A little less improvement than GuestOS makes much more sense. 👍

**Pull request merged**